test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableTestCase.java

```diff
@@ -0,0 +1,98 @@
+            assertFalse("query is equal to null", firstInstance.equals(null));
+            assertFalse("query is equal to incompatible type", firstInstance.equals(""));
+            assertTrue("query is not equal to self", firstInstance.equals(firstInstance));
+            assertThat("same query's hashcode returns different values if called multiple times", firstInstance.hashCode(),
+                    equalTo(firstInstance.hashCode()));
+
+            T secondInstance = copyInstance(firstInstance);
+            assertTrue("query is not equal to self", secondInstance.equals(secondInstance));
+            assertTrue("query is not equal to its copy", firstInstance.equals(secondInstance));
+            assertTrue("equals is not symmetric", secondInstance.equals(firstInstance));
+            assertThat("query copy's hashcode is different from original hashcode", secondInstance.hashCode(),
+                    equalTo(firstInstance.hashCode()));
+
+            T thirdInstance = copyInstance(secondInstance);
+            assertTrue("query is not equal to self", thirdInstance.equals(thirdInstance));
+            assertTrue("query is not equal to its copy", secondInstance.equals(thirdInstance));
+            assertThat("query copy's hashcode is different from original hashcode", secondInstance.hashCode(),
+                    equalTo(thirdInstance.hashCode()));
+            assertTrue("equals is not transitive", firstInstance.equals(thirdInstance));
+            assertThat("query copy's hashcode is different from original hashcode", firstInstance.hashCode(),
+                    equalTo(thirdInstance.hashCode()));
+            assertTrue("equals is not symmetric", thirdInstance.equals(secondInstance));
+            assertTrue("equals is not symmetric", thirdInstance.equals(firstInstance));
+        }
+    }
+
+    /**
+     * Test serialization and deserialization of the test query.
+     */
+    public void testSerialization() throws IOException {
+        for (int runs = 0; runs < NUMBER_OF_TESTQUERIES; runs++) {
+            T testInstance = createTestInstance();
+            assertSerialization(testInstance);
+        }
+    }
+
+    /**
+     * Serialize the given query builder and asserts that both are equal
+     */
+    protected T assertSerialization(T testInstance) throws IOException {
+        T deserializedInstance = copyInstance(testInstance);
+        assertEquals(testInstance, deserializedInstance);
+        assertEquals(testInstance.hashCode(), deserializedInstance.hashCode());
+        assertNotSame(testInstance, deserializedInstance);
+        return deserializedInstance;
+    }
+
+    private T copyInstance(T instance) throws IOException {
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            instance.writeTo(output);
+            try (StreamInput in = new NamedWriteableAwareStreamInput(output.bytes().streamInput(),
+                    new NamedWriteableRegistry(Collections.emptyList()))) {
```

Comment:
will this work for all serialization classes? I mean queries need to be registered etc. I think we have to add all the ones we need for the subclass. So maybe add a factory method we can override?